### DISCUSSION TITLE
Can you try Ghidra's debugger? Having problems with it.

### DISCUSSION THREAD

**Discussion by SirEnder125** [RedShyGuy/Vapecord-ACNL-Plugin#82] [General]
Hey @RedShyGuy. I've lately been trying to come up with a cheat or two for ACNL. I know you use IDA Pro, but I was wondering if you might be willing to try out Ghidra's debugger. Ghidra's developers were trying to help me use it on my 3ds, but I got some weird GDB errors. So I was wondering if it was just me. Please consider trying it out when you have some time. :D Here's a link to a conversation (which is rather long) trying to use Ghidra's debugger for a 3ds.

[conversation](https://github.com/NationalSecurityAgency/ghidra/discussions/2721)

**Comment by RedShyGuy** [comment]
Hey! I might be able to look into it yea

**Reply by SirEnder125** [reply]
In [Ghidra's github discussions page,](https://github.com/NationalSecurityAgency/ghidra/discussions?discussions_q=author%3ASirEnder125) I started a few threads asking for help. In Ghidra's debugger, I have to connect via SSH to WSL (Windows Subsystem for Linux), then Ghidra's debugger launches gdb on WSL, then I connect to Luma3ds via the interpreter. It's a lot of work as far as I know ,but I thought you may be familiar with it, leastwise more than I.

**Comment by SirEnder125** [comment]
Hey, @RedShyGuy. With [a ton of help from the developers of Ghidra,](https://github.com/NationalSecurityAgency/ghidra/discussions/3590) I have Ghidra connected to Luma3ds, and I can set watchpoints, and edit bytes in the dynamic listing. But I am a little rusty when it comes to making cheats, especially when it comes to using a debugger. So could you maybe tell me how to make a small beginner cheat using a debugger? For instance, how did you originally find the address you told me to hook on in #41? I believe it was `0x2C02E0`, but when I set a watch on the bank amount address, I got `0x3036f8`. 

I don't know how different Ghidra is from IDA Pro, but I would assume it's not too drastic of a change.

**Reply by RedShyGuy** [reply]
Glad you got it working!
And the function you got there is the general encryption function. It is used for money so it is correct, but it isn't only used for money. So the function I showed you is the one only used for money.
To find it you simply need to look for XREFs of the function, by for example putting a breakpoint on the function you found, then look at LR, it will most of the time hold the function which called the function you put the breakpoint on.

**Reply by SirEnder125** [reply]
Alright. To be sure, that's the register LR you are referring to? I'm not sure I quite understand, but here's a image of the currently disassembled part of the function (I think?):
![image](https://user-images.githubusercontent.com/69535694/141248688-0e18a6c8-9724-4991-831b-0af289ac3070.png)

I set a breakpoint on 0x3036F8, as seen in the next image:
![image](https://user-images.githubusercontent.com/69535694/141248867-d1835a11-2e06-413c-a8c4-e5314ff49e8a.png)

I then resumed execution and changed the amount of money in my wallet. The breakpoint was hit, and the LR register was 0x3036ec (just a few addresses before the one that the breakpoint was set on. (This seems wrong...)

Should the LR register have the value 0x2C02E0? Or do I have to back-trace further?

**Reply by RedShyGuy** [reply]
You must set the breakpoint at the function start. The function starts at `0x3036A4`, set a breakpoint onto it

**Reply by SirEnder125** [reply]
Okay, cool. Out of curiosity, how do you know where a function starts?

**Reply by RedShyGuy** [reply]
Many functions start with register storing instructions, but not all. The easiest way to know where they start is by using Ghidra and analyzing the whole .elf of the game, it will try to find all the function beginnings, and it does a good job on it

**Reply by SirEnder125** [reply]
Hmm. Alright. Sounds complicated, I know a .elf file is a few steps below an executable. Do you just search the memory, is the .elf data in the ram, or do you have to have the actual file? If so, how? (Sorry, I understand this may be a little off-topic as far as this discussion is concerned.)

**Reply by RedShyGuy** [reply]
The .elf is a file. You can create it by using a special script. You just need the code.bin and the exheader of the game (you get those 2 from GodMode9 for example)

**Reply by SirEnder125** [reply]
Ah.

**Reply by SirEnder125** [reply]
~~I'm not sure I'd know how to use the .elf file if I had it anyway, but I need something to help with what I am doing. Do you know the name of the script? I have Discord, if it's confidential in some way.~~

Nevermind, I got the elf file. Do you just open it in the listing window of the debugger? Are all these FUN_* functions real functions in ACNL?

**Reply by RedShyGuy** [reply]
Hey man if you have Discord, please let us switch to it. It is way easier to chat there

**Reply by SirEnder125** [reply]
Alright. 👍🏻

**Reply by SirEnder125** [reply]
Lukas#4444, is it?
